### PR TITLE
POL-38 - Google cloud cud report

### PR DESCRIPTION
**Pull request by nshivappa**
### Description

This policy generates the Report on Google-cloud Committed Use discounts(CUD) which has a user input email, google_project ID, CUD status(ALL, Active, Expired).

### Prerequisite to apply this policy

- Add New credentials (GC_SA_CLIENT_EMAIL and GC_SA_PRIVATE_KEY) *if does not exists* under RightScale cloud management Desing/Credentials. 
- The value for credentials can be found in IAM & admin\service accounts under Google-cloud platform.

Note: The Service Account in GCP should have *owner role*

### Issues Resolved

Initial version: 1.0

### Contribution Check List

- [ ] All tests pass.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable
- [ ] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  label "Email addresses of the recipients you wish to notify"
+end
+
+parameter "param_google_project" do
+  type "string"
+  label "Google Cloud Project"
+  allowed_pattern /^[0-9a-z:\.-]+$/
+end
+
+parameter "param_google_status" do
```
Comment (ryanoleary) [inline cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: maybe name param_cud_status - it's the status of the CUD, not the google project

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+ label "CUD status"
+ type "string"
+ allowed_values "ALL","ACTIVE","EXPIRED"
```
Comment (ryanoleary) [inline cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: For readability, might consider calling these "All", "Active", and "Expired". Can have a mapping later that maps those to the actual API values to use

File: cost/google/cud_report/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ 
+- Google
+ 
+### Cost
+ 
+This Policy Template does not incur any cloud costs.
+
+### Prerequisite to apply this policy
+
+- Add New credentials (GC_SA_CLIENT_EMAIL and GC_SA_PRIVATE_KEY) *if doesnot exists* under Rightscale cloud managemant Desing/Credentials.
```
Comment (ryanoleary) [inline cost/google/cud_report/README.md]: I don't see either of these credentials used in the policy.

What would the email cred be used for, if it was used?

Check spelling in this sentence.

Change Rightscale to RightScale

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+auth "auth_google", type: "oauth2" do
+  token_url "https://oauth2.googleapis.com/token"
+  grant type: "jwt_bearer" do
+  iss cred("GCE_PLUGIN_ACCOUNT")
+  aud "https://oauth2.googleapis.com/token"
+  additional_claims do {
+      "scope" => "https://www.googleapis.com/auth/compute"
+    } end
+  signing_key cred("GCE_PLUGIN_PRIVATE_KEY")
```
Comment (ryanoleary) [inline cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: Recommend remove PLUGIN from this cred name

File: cost/google/cud_report/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
+## Google Committed Use Discount (CUD) Report
+ 
+### What it does
+This policy identifies all CUDs that exist in a given GCP project and provides a report listing them all. It can optionally report on all CUDs, or only those that are active or expired.
+ 
+### Functional Details
+ 
+Uses the GCP API to get a list of all CUDs and report on them.
```
Comment (ryanoleary) [inline cost/google/cud_report/README.md]: I would add some detail here about using a service account and what perms are required for that.

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt (hunk 1 above)
Comment (nshivappa) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: Changed "param_google_status" to "param_cud_status"

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt (hunk 2 above)
Comment (nshivappa) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: Made required changes. now it displays "All", "Active", and "Expired" in a dropdown and converting to uppercase while comparing.

File: cost/google/cud_report/README.md (hunk 3 above)
Comment (nshivappa) [reply cost/google/cud_report/README.md]: I have renamed "GCE_PLUGIN_ACCOUNT and GCE_PLUGIN_PRIVATE_KEY" to "GC_SA_CLIENT_EMAIL and GC_SA_PRIVATE_KEY" respectively.

 'email cred' is a client email of the particular service account, which is used for authentication.

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt (hunk 4 above)
Comment (nshivappa) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: Renamed "GCE_PLUGIN_ACCOUNT and GCE_PLUGIN_PRIVATE_KEY" with "GC_SA_CLIENT_EMAIL and GC_SA_PRIVATE_KEY" respectively

File: cost/google/cud_report/README.md (hunk 5 above)
Comment (nshivappa) [reply cost/google/cud_report/README.md]: Updated file with information on creating a Service account.

**Comment by nshivappa** [comment]
Please find below link where Template is working.
https://governance.rightscale.com/org/1105/projects/60073/applied-policies

**Comment by ryanoleary** [comment]
Looks great.

It would be good if the title and/or table header in the incident contained the project ID against which this policy was run, since that is not part of the returned data.

File: cost/google/cud_report/README.md [hunk 6]
```diff
@@ -26,7 +27,7 @@ This Policy Template does not incur any cloud costs.
 
 ### Prerequisite to apply this policy
 
-- Add New credentials (GC_SA_CLIENT_EMAIL and GC_SA_PRIVATE_KEY) *if doesnot exists* under Rightscale cloud managemant Desing/Credentials. 
-- The value for credentials can be found in IAM & admin\service accounts under Google-cloud platform.
+- Add New credentials (GC_SA_CLIENT_EMAIL and GC_SA_PRIVATE_KEY) *if does not exists* under RightScale cloud management Desing/Credentials.
```
Comment (ryanoleary) [inline cost/google/cud_report/README.md]: Spelling on "Desing"

**Comment by ryanoleary** [comment]
Other than my final comments it looks good. 

It looks like there are still some other files in this PR which shouldn't be there, but someone on the team (@douglaswth ?) that can advise on that.

**Comment by nshivappa** [comment]
> Looks great.
> 
> It would be good if the title and/or table header in the incident contained the project ID against which this policy was run, since that is not part of the returned data.

Included Project ID in the title.
Now the title looks like below
List of Committed Use Discounts in Google cloud for Project: “<<----Project ID---->>”

File: cost/google/cud_report/README.md (hunk 6 above)
Comment (nshivappa) [reply cost/google/cud_report/README.md]: I have updated the file.

**Comment by srpomeroy** [comment]
Internal ticket: POL-38

File: cost/google/cud_report/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+## Google Committed Use Discount (CUD) Report
+
```
Comment (srpomeroy) [inline cost/google/cud_report/README.md]: Need to add a note similar to the other RightScale account independent policies.
https://github.com/rightscale/policy_templates/tree/master/cost/budget_alerts

**As a best practice, this policy should only be applied to the Master Account, and not to each individual RightScale Account.**

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  }
+EOS
+end
+
+
+escalation "report_list_of_CUDs" do
+   email $param_email
+end
+
+policy "policy_committed_use_discount" do
```
Comment (srpomeroy) [inline cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: Recommend we update the wording in the policy template piece to remove RS Account from the subject as its not pertinent to the incident data. By moving it to the bottom, it still allows the person receiving the email to know where the policy is applied without being the focus of the email. Also added some additional tweaks by adding the project and the status type to the subject.

Something like this:
```
policy "policy_committed_use_discount" do
  validate $ds_committed_use_discount_map do
    summary_template "{{ parameters.param_google_project }} : List of {{ parameters.param_cud_status }} Committed Use Discounts in Google Cloud"
    detail_template <<-EOS
# {{ parameters.param_cud_status }} Committed Use Discounts in Google Cloud for Project: {{ parameters.param_google_project }}

| CUD Id |  Name  | Region | Status | Plan | Start Timestamp | End Timestamp |
| ------ | ------ | ------ | ------ | ---- | --------------- | ------------- |
{{ range data -}}
| {{.cud_id}} | {{.name}} | {{.region}} | {{.status}} | {{.plan}} | {{.startTimestamp}} | {{.endTimestamp}} |
{{ end -}}

___
###### Policy Applied in Account: {{ rs_project_name }} (Account ID: {{ rs_project_id }}) within Org: {{ rs_org_name }} (Org ID: {{ rs_org_id }})
EOS
  escalate $report_list_of_CUDs
  check eq(size(data),0)
  end
end```

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt [hunk 9]
```diff
@@ -0,0 +1,134 @@
+name "Google Committed Use Discount (CUD) Report"
+rs_pt_ver 20180301
+type "policy"
+short_description "Report on all Google CUDs that have been purchased. \n See the [README](https://github.com/rightscale/policy_templates/tree/master/cost/google/cud_report) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more."
```
Comment (srpomeroy) [inline cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: In an attempt to educate the user, recommend adding an additional line to the short description in line with the note in the readme about best practices in applying the policy.

>short_description "Report on all Google CUDs that have been purchased. See the [README](https://github.com/rightscale/policy_templates/tree/master/cost/google/cud_report) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more. \n
>As a best practice, this policy should only be applied to the Master Account, and not to each individual RightScale Account."

**Comment by srpomeroy** [comment]
Also need to add the policy to the root readme: https://github.com/rightscale/policy_templates/blob/master/README.md

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+datasource "ds_committed_use_discount_map" do
+  run_script $js_committed_use_discount_map, $ds_committed_use_discount, $param_cud_status
+end
+
+#Filter the JSON, based on user selection for param 'param_cud_status' and formating Date.
+script "js_committed_use_discount_map", type: "javascript" do
+  parameters "commitments","param_cud_status"
+  result "content"
+  code <<-EOS
+  console.log(commitments)
```
Comment (cdwilhelm) [inline cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: When done debugging remove the console.log statements

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt (hunk 8 above)
Comment (srpomeroy) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: @ryanoleary seeing as you just asked for this to be added, would like your feedback/ok here.

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt (hunk 9 above)
Comment (srpomeroy) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: @ryanoleary Should we specify "single account" as opposed to "master account"?

File: cost/google/cud_report/README.md (hunk 7 above)
Comment (srpomeroy) [reply cost/google/cud_report/README.md]: @ryanoleary Should we specify "single account" as opposed to "master account"?

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt (hunk 8 above)
Comment (srpomeroy) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: I talked to @ryanoleary about this, and he agrees. Google project should be in the `summary_template` and the `detail_template`.

Comment (nshivappa) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: @srpomeroy  I have made the required changes.

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt (hunk 10 above)
Comment (nshivappa) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: Thanks, @cdwilhelm  I Removed the console.log.

**Comment by nshivappa** [comment]
> Also need to add the policy to the root readme: https://github.com/rightscale/policy_templates/blob/master/README.md

I think once the code is merged into the master branch, we can add it in root README.md

**Comment by cdwilhelm** [comment]
@nshivappa you include  the ROOT readme in this PR.  There is no need to wait. 

> > Also need to add the policy to the root readme: https://github.com/rightscale/policy_templates/blob/master/README.md
> 
> I think once the code is merged into the master branch, we can add it in root README.md

@nshivappa you don't need to wait, you can include the root READMe in this PR too.

**Comment by srpomeroy** [comment]
Example of most recent version applied: https://governance.rightscale.com/org/1105/projects/60073/applied-policies/5c7fdc050aa25c00889b3e51

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+          }
+
+  if(param_cud_status != 'All'){
+    if(commitments[i].status === param_cud_status.toUpperCase()){
+    content.push(commitmentDetails)
+    }
+  }else{
+    content.push(commitmentDetails)
+  }
+  }
```
Comment (douglaswth) [inline cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: The formatting of this JavaScript code could be formatted better, the indentation seems off.

**Comment by nshivappa** [comment]
> @nshivappa Do you have the update Policy Template running somewhere?

https://governance.rightscale.com/org/1105/projects/60073/applied-policies/5c80e3107dc2d00088ff93de

File: cost/google/cud_report/Google_Committed_UseDiscount_Report.pt (hunk 11 above)
Comment (nshivappa) [reply cost/google/cud_report/Google_Committed_UseDiscount_Report.pt]: I have Formated the JavaScript code.

**Comment by nshivappa** [comment]
> I think some of the markdown files that were restored may have previously had inconsistent line endings and that may be why they are still showing up in the diff. It might be a good idea to add some settings to make sure line endings are consistent in the `.gitattributes` file.



> It might be a good idea to add some settings to make sure line endings are consistent in the `.gitattributes` file.

As I am new for these things, I am not sure what changes has to be done in `.gitattributes` file.

**Comment by douglaswth** [comment]
This is what we use in other repositories:

```
# Force Unix line endings even on Windows
* text=auto eol=lf
```

**Pull request merged**